rfc: don't panic (#7472)
tendermint/tendermint: docs/rfc/rfc-008-don't-panic.md

where it is acceptable to panic.

### References

- [Defer Panic and Recover](https://go.dev/blog/defer-panic-and-recover)
- [Why Go gets exceptions right](https://dave.cheney.net/tag/panic)
- [Don't panic](https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_dont_panic)

## Discussion

### Acceptable Panics

#### Initialization

It is unambiguously safe (and desireable) to panic in `init()`
functions in response to any kind of error. These errors are caught by
tests, and occur early enough in process initialization that they
won't cause unexpected runtime crashes.

Other code that is called early in process initialization MAY panic,
in some situations if it's not possible to return an error or cause
the process to abort early, although these situations should be
vanishingly slim.

#### Data Corruption

If Tendermint code encounters an inconsistency that could be
attributed to data corruption or a logical impossibility it is safer
to panic and crash the process than continue to attempt to make
progress in these situations.

Examples including reading data out of the storage engine that
is invalid or corrupt, or encountering an ambiguous situation where

> the process should halt. Generally these forms of corruption are

Reviewer: I'm wondering if you think that this should include potentially ambiguous cases in the consensus logic. 
An example is this line of code
https://github.com/tendermint/tendermint/blob/ab7da86b06233b325406d8b0fb5ff0902adec001/internal/consensus/state.go#L1511

We panic here because locally, we think that the block we are looking at is invalid, but for some reason a large part of the network did not think that it was invalid. We likely want to shutdown the process because it's likely that the program can no longer make meaningful progress: it can't validate future blocks if the network is behaving in a way it doesn't understand etc. Therefore, we likely want to stop, but I'm not certain if panicking makes sense in this case?

Reviewer 2: Yes, I think the important "test" is whether anything meaningful can be done with an error.  

"Meaningful" could be as simple as logging, failing a transaction, and moving on. But if continuing would leave the system in an incoherent state (which I think is what you're describing) it would be better to terminate the program. In those cases a panic is fine, and probably preferable to an explicit exit, since it lets you shut down cleanly and also write tests that exercise the condition.

Author: I think this is a good place to panic based on your description of what is happening here.

Reviewer: I think it'd be worth adding that somewhat explicitly to this document. Something along the lines of

```
In general, a panic should be used in the case of unrecoverable consensus failures. If a process detects that the network is behaving in an incoherent way and it does not have a clearly defined and mechanism for recovering, the process should panic.
```

detected after interacting with a trusted but external data source,
and reflect situations where the author thinks its safer to terminate
the process immediately rather than allow execution to continue.

TODO: Add more and better examples here.

#### Static Validity

It is acceptable to panic for invariant violations, within a library
or package, in situations that should be statically impossible,
because there is no way to make these kinds of assertions at compile
time.

For example, type-asserting `interface{}` values returned by
`container/list` and `container/heap` (and similar), is acceptable,
because package authors should have exclusive control of the inputs to
these containers. Packages should not expose the ability to add
arbitrary values to these data structures.

#### Controlled Panics Within Libraries

In some algorithms with highly recursive structures or very nested
call patterns, using a panic, in combination with conditional recovery
handlers results in more manageable code. Ultimately this is a limited
application, and implementations that use panics internally should
only recover conditionally, filtering out panics rather than ignoring

> or handling all panics.

Reviewer: I would be interested to see some more examples of this. To my mind, a panic is generally for when something is critically and unrecoverably wrong. The code cannot do its job and may need intervention before starting up again. If the issue can be handled, why not just return an error?

#### Request Handling

Code that handles responses to incoming/external requests
(e.g. `http.Handler`) should avoid panics, but practice this isn't
totally possible, and it makes sense that request handlers have some
kind of default recovery mechanism that will prevent one request from
terminating a service.

### Unacceptable Panics

In **no** other situation is it acceptable for the code to panic:

> - there should be controlled panics that should be handled across

Reviewer: I'm a little unclear, are you saying that this is a case where panics should not exist? I generally think that we should endeavor to avoid panics in this case and not ever expect to handle a panic for a package we called into. If the package is expected to shutdown in error, it should communicate that with an error.

Author: I think I forgot some negation words here. I've cleaned that up.

  library/package boundaries.
- callers of library functions should not expect panics.
- ensuring that arbitrary go routines can't panic.
- ensuring that there are no arbitrary panics in core production code,

>   espically code that can run at any time during the.

Reviewer: accidentally a word?

- all test code and fixture should report normal test assertions with
  a mechanism like testify's `require` assertion rather than calling
  panic directly.

The goal of this increased "panic rigor" is to ensure that any escaped
panic is reflects a fixable bug in Tendermint.

### Removing Panics

The process for removing panics involve a few steps, and will be part
of an ongoing process of code modernization:

> - converting existing explicit panics to errors in cases where it's
>   possible to return an error, the errors would not be ignored, and
>   not panicing would not lead to data corruption or cover up data
>   corruption.

Reviewer: ```suggestion
- converting existing explicit panics to errors in cases where it's
  possible to return an error, the errors can and should be handled, and returning
  an error would not lead to data corruption or cover up data
  corruption.
```

Is this roughly what's beeing said here?

because there is no way to make these kinds of assertions at compile
time.

For example, type-asserting `interface{}` values returned by
`container/list` and `container/heap` (and similar), is acceptable,
because package authors should have exclusive control of the inputs to
these containers. Packages should not expose the ability to add
arbitrary values to these data structures.

#### Controlled Panics Within Libraries

In some algorithms with highly recursive structures or very nested
call patterns, using a panic, in combination with conditional recovery
handlers results in more manageable code. Ultimately this is a limited
application, and implementations that use panics internally should
only recover conditionally, filtering out panics rather than ignoring
or handling all panics.

#### Request Handling

Code that handles responses to incoming/external requests
(e.g. `http.Handler`) should avoid panics, but practice this isn't
totally possible, and it makes sense that request handlers have some
kind of default recovery mechanism that will prevent one request from
terminating a service.

### Unacceptable Panics

In **no** other situation is it acceptable for the code to panic:

> - there should be controlled **no** panics that callers are required

Reviewer: ```suggestion
- there should be **no** controlled panics that callers are required
```
